Allow constructing a PailSource from a PailSpec

amazari opened the pull request:
By allowing a PailSource to be constructed from a PailSpec, this patch gives more flexibility to client code. For example, it is is now possible to set compression and/or sizing policy.
For backward compatibility sake, it is still possible to construct a source from a PailStructure. In that case, a PailSpec is materialized using PailTap.makeSpec facility in the same manner it was previously done.

Commit:
Add a PailSource ctor taking a PailSpec arg

Exposing a PailSpec-taking ctor give more flexibility in PailSource
construction. For example, compression and/or sizing policy can be
set.

Commit:
Merge remote-tracking branch 'upstream/develop' into develop

ianoc commented on scalding-commons/src/main/scala/com/twitter/scalding/commons/source/PailSource.scala:
```diff
@@ -153,7 +169,6 @@ class PailSource[T] private (rootPath: String, structure: PailStructure[T], subP
 
   lazy val getTap = {
-    val spec = PailTap.makeSpec(null, structure)
     val javaSubPath = if ((subPaths == null) || (subPaths.size == 0)) null else subPaths map { _.asJava }
```
Since the constructor setting up the null for the list is private, and we only seem to care if the path is empty or not. Can we just default subPaths to an empty list and not need to do null checking?

Commit:
Merge remote-tracking branch 'upstream/develop' into develop

Commit:
Use an empty array as default value for subpath

Assertions in factory functions and sane default value
for the subPath field of PailSource remove the need for
null-checking the val when creating a Tap.

johnynek commented on scalding-commons/src/main/scala/com/twitter/scalding/commons/source/PailSource.scala:
```diff
@@ -145,7 +161,7 @@ object PailSource {
 
-class PailSource[T] private (rootPath: String, structure: PailStructure[T], subPaths: Array[List[String]] = null)(implicit conv: TupleConverter[T])
+class PailSource[T] private (rootPath: String, spec: PailSpec, subPaths: Array[List[String]] = Array.empty)(implicit conv: TupleConverter[T])
   extends Source with Mappable[T] {
```
can you add with `TypedSink[T]`? Is there any reason why you can't write to this?

Commit:
Merge remote-tracking branch 'upstream/develop' into develop

Commit:
Make PailSource a TypeSink

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/twitter/scalding?pullRequest=1424) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/twitter/scalding?pullRequest=1424) before we can accept your contribution.<br/><hr/>**Alexandre Mazari** seems not to be a GitHub user. You need a GitHub account to be able to sign the CLA. If you have already a GitHub account, please [add the email address used for this commit to your account](https://help.github.com/articles/why-are-my-commits-linked-to-the-wrong-user/#commits-are-not-linked-to-any-user).<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/twitter/scalding?pullRequest=1424) it.</sub>
